src/libsyntax/parse/parser.rs

```diff
@@ -7523,10 +7523,26 @@ impl<'a> Parser<'a> {
         if !any_disr.is_empty() && !all_nullary {
             let mut err = self.struct_span_err(
                 any_disr.clone(),
-                "discriminator values can only be used with a field-less enum",
+                "custom discriminant values are not allowed in enums with fields",
             );
             for sp in any_disr {
-                err.span_label(sp, "only valid in field-less enums");
+                err.span_label(sp, "invalid custom discriminant");
+            }
+            for variant in &variants {
+                if let VariantData::Struct(fields, ..) | VariantData::Tuple(fields, ..) =
+                    &variant.node.data
+                {
+                    let fields = if fields.len() > 1 {
```

Comment:
@estebank Feels like we should have a general framework for these sorts of things...? `trait Article`?